src/rsa/bigint.rs

```diff
@@ -72,21 +110,217 @@ impl Positive {
+    pub fn try_clone(&self) -> Result<OddPositive, error::Unspecified> {
+        let mut value = try!(Whole::zero());
+        try!(bssl::map_result(unsafe {
+            GFp_BN_copy(value.as_mut_ref(), self.as_ref())
+        }));
+        Ok(OddPositive(Positive(value)))
+    }
+
+    pub fn into_elem<F: Field>(self, m: &Modulus<F>)
+            -> Result<Elem<F>, error::Unspecified> {
+        self.0.into_elem(m)
+    }
+
+    pub fn into_elem_decoded<F: Field>(self, m: &Modulus<F>)
+            -> Result<ElemDecoded<F>, error::Unspecified> {
+        self.0.into_elem_decoded(m)
+    }
+
+    pub fn into_modulus<F: Field>(self)
+                                  -> Result<Modulus<F>, error::Unspecified> {
+        let r = Modulus {
+            ctx: unsafe { GFp_BN_MONT_CTX_new() },
+            field: PhantomData,
+        };
+        if r.ctx.is_null() {
+            return Err(error::Unspecified);
+        }
+        // XXX: This makes a copy of `self`'s `BIGNUM`. TODO: change this to a
+        // move.
+        try!(bssl::map_result(unsafe {
+            GFp_BN_MONT_CTX_set(&mut *r.ctx, self.as_ref())
+        }));
+        Ok(r)
+    }
 }
 
+impl core::ops::Deref for OddPositive {
+    type Target = Positive;
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+
+/// Every modulus (`n`, `p`, `q`, etc.) should be represented as a separate
+/// type implementing `Field`.
+pub unsafe trait Field {}
+
+/// A modulus that can be used for Montgomery math.
+pub struct Modulus<F: Field> {
+    ctx: *mut BN_MONT_CTX,
+    field: PhantomData<F>,
+}
+
+impl<F: Field> Modulus<F> {
+    pub fn as_ref(&self) -> &BN_MONT_CTX { unsafe { &*self.ctx } }
+}
+
+impl<F: Field> Drop for Modulus<F> {
+    fn drop(&mut self) { unsafe { GFp_BN_MONT_CTX_free(self.ctx); } }
+}
+
+/// Montgomery-encoded elements of a field.
+pub struct Elem<F: Field> {
+    value: Whole,
+    field: PhantomData<F>,
+}
+
+impl<F: Field> Elem<F> {
+    pub fn as_ref_montgomery_encoded<'a>(&'a self) -> &'a BIGNUM {
+        self.value.as_ref()
+    }
+}
+
+pub struct ElemDecoded<F: Field> {
+    value: Whole,
+    field: PhantomData<F>
+}
+
+impl<F: Field> ElemDecoded<F> {
+    pub fn fill_be_bytes(&self, out: &mut [u8])
+                         -> Result<(), error::Unspecified> {
+        bssl::map_result(unsafe {
+            GFp_BN_bn2bin_padded(out.as_mut_ptr(), out.len(),
+                                 self.value.as_ref())
+        })
+    }
+
+    pub fn is_zero(&self) -> bool { self.value.is_zero() }
+
+    pub fn is_one(&self) -> bool { self.value.is_one() }
+
+    // XXX: This makes it too easy to break the invariants. TOOD: Remove this
+    // ASAP.
+    pub unsafe fn as_mut_ref<'a>(&'a mut self) -> &'a mut BIGNUM {
+        self.value.as_mut_ref()
+    }
+
+    pub fn into_odd_positive(self) -> Result<OddPositive, error::Unspecified> {
+        self.value.into_odd_positive()
+    }
+}
+
+// `a` * `b` (mod `n`), assuming `a` and `b` are both less than `m`.
```

Comment:
I also removed the "assuming.." {a, b} <= m is implied by the use of the `Elem[Decoded]` types.